File: concepts/03-stacks/node/stack.js
/**
 * Stack Implementation in Node.js
 */

class Stack {
    /**
     * Array-based Stack Implementation
     */
    constructor() {
        this.items = [];
    }

    /**
     * Add element to top - O(1)
     */
    push(item) {
        this.items.push(item);
    }

    /**
     * Remove and return top element - O(1)
     */
    pop() {
        if (this.isEmpty()) {
            throw new Error("Stack is empty");
        }
        return this.items.pop();
    }

    /**
     * Return top element without removing - O(1)
     */
    peek() {
        if (this.isEmpty()) {
            throw new Error("Stack is empty");
        }
        return this.items[this.items.length - 1];
    }

    /**
     * Check if stack is empty - O(1)
     */
    isEmpty() {
        return this.items.length === 0;
    }

    /**
     * Return size of stack - O(1)
     */
    size() {
        return this.items.length;
    }

    /**
     * Display all elements
     */
    display() {
        return [...this.items];
    }
}

class StackNode {
    /**
     * Node for linked list-based stack
     */
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedStack {
    /**
     * Linked List-based Stack Implementation
     */
    constructor() {
        this.top = null;
        this.size = 0;
    }

    /**
     * Add element to top - O(1)
     */
    push(item) {
        const newNode = new StackNode(item);
        newNode.next = this.top;
        this.top = newNode;
        this.size++;
    }

    /**
     * Remove and return top element - O(1)
     */
    pop() {
        if (this.isEmpty()) {
            throw new Error("Stack is empty");
        }
        const data = this.top.data;
        this.top = this.top.next;
        this.size--;
        return data;
    }

    /**
     * Return top element without removing - O(1)
     */
    peek() {
        if (this.isEmpty()) {
            throw new Error("Stack is empty");
        }
        return this.top.data;
    }

    /**
     * Check if stack is empty - O(1)
     */
    isEmpty() {
        return this.top === null;
    }

    /**
     * Return size of stack - O(1)
     */
    size() {
        return this.size;
    }
}

class MinStack {
    /**
     * Stack that supports getMin() in O(1)
     */
    constructor() {
        this.stack = [];
        this.minStack = [];
    }

    /**
     * Push element - O(1)
     */
    push(val) {
        this.stack.push(val);
        if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1]) {
            this.minStack.push(val);
        }
    }

    /**
     * Pop element - O(1)
     */
    pop() {
        if (this.stack.length === 0) {
            throw new Error("Stack is empty");
        }
        const val = this.stack.pop();
        if (val === this.minStack[this.minStack.length - 1]) {
            this.minStack.pop();
        }
        return val;
    }

    /**
     * Get top element - O(1)
     */
    top() {
        if (this.stack.length === 0) {
            throw new Error("Stack is empty");
        }
        return this.stack[this.stack.length - 1];
    }

    /**
     * Get minimum element - O(1)
     */
    getMin() {
        if (this.minStack.length === 0) {
            throw new Error("Stack is empty");
        }
        return this.minStack[this.minStack.length - 1];
    }
}

// Example usage
if (require.main === module) {
    // Array-based Stack
    console.log("--- Array-based Stack ---");
    const stack = new Stack();
    stack.push(1);
    stack.push(2);
    stack.push(3);
    console.log("Stack:", stack.display());
    console.log("Top:", stack.peek());
    console.log("Pop:", stack.pop());
    console.log("After pop:", stack.display());

    // Linked Stack
    console.log("\n--- Linked Stack ---");
    const linkedStack = new LinkedStack();
    linkedStack.push(10);
    linkedStack.push(20);
    linkedStack.push(30);
    console.log("Size:", linkedStack.size());
    console.log("Top:", linkedStack.peek());
    console.log("Pop:", linkedStack.pop());

    // Min Stack
    console.log("\n--- Min Stack ---");
    const minStack = new MinStack();
    minStack.push(3);
    minStack.push(1);
    minStack.push(5);
    console.log("Min:", minStack.getMin());
    minStack.pop();
    console.log("After pop, Min:", minStack.getMin());
}

module.exports = { Stack, LinkedStack, MinStack };

File: concepts/03-stacks/python/stack.py
"""
Stack Implementation in Python
"""

class Stack:
    """Array-based Stack Implementation"""
    
    def __init__(self):
        self.items = []
    
    def push(self, item):
        """Add element to top - O(1)"""
        self.items.append(item)
    
    def pop(self):
        """Remove and return top element - O(1)"""
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.items.pop()
    
    def peek(self):
        """Return top element without removing - O(1)"""
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.items[-1]
    
    def is_empty(self):
        """Check if stack is empty - O(1)"""
        return len(self.items) == 0
    
    def size(self):
        """Return size of stack - O(1)"""
        return len(self.items)
    
    def display(self):
        """Display all elements"""
        return self.items.copy()


class StackNode:
    """Node for linked list-based stack"""
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedStack:
    """Linked List-based Stack Implementation"""
    
    def __init__(self):
        self.top = None
        self.size = 0
    
    def push(self, item):
        """Add element to top - O(1)"""
        new_node = StackNode(item)
        new_node.next = self.top
        self.top = new_node
        self.size += 1
    
    def pop(self):
        """Remove and return top element - O(1)"""
        if self.is_empty():
            raise IndexError("Stack is empty")
        data = self.top.data
        self.top = self.top.next
        self.size -= 1
        return data
    
    def peek(self):
        """Return top element without removing - O(1)"""
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.top.data
    
    def is_empty(self):
        """Check if stack is empty - O(1)"""
        return self.top is None
    
    def size(self):
        """Return size of stack - O(1)"""
        return self.size


class MinStack:
    """Stack that supports getMin() in O(1)"""
    
    def __init__(self):
        self.stack = []
        self.min_stack = []
    
    def push(self, val):
        """Push element - O(1)"""
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
    
    def pop(self):
        """Pop element - O(1)"""
        if not self.stack:
            raise IndexError("Stack is empty")
        val = self.stack.pop()
        if val == self.min_stack[-1]:
            self.min_stack.pop()
        return val
    
    def top(self):
        """Get top element - O(1)"""
        if not self.stack:
            raise IndexError("Stack is empty")
        return self.stack[-1]
    
    def get_min(self):
        """Get minimum element - O(1)"""
        if not self.min_stack:
            raise IndexError("Stack is empty")
        return self.min_stack[-1]


# Example usage
if __name__ == "__main__":
    # Array-based Stack
    print("--- Array-based Stack ---")
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print("Stack:", stack.display())
    print("Top:", stack.peek())
    print("Pop:", stack.pop())
    print("After pop:", stack.display())
    
    # Linked Stack
    print("\n--- Linked Stack ---")
    linked_stack = LinkedStack()
    linked_stack.push(10)
    linked_stack.push(20)
    linked_stack.push(30)
    print("Size:", linked_stack.size())
    print("Top:", linked_stack.peek())
    print("Pop:", linked_stack.pop())
    
    # Min Stack
    print("\n--- Min Stack ---")
    min_stack = MinStack()
    min_stack.push(3)
    min_stack.push(1)
    min_stack.push(5)
    print("Min:", min_stack.get_min())
    min_stack.pop()
    print("After pop, Min:", min_stack.get_min())

